### PR TITLE
Adding credential helper

### PR DESCRIPTION
**Pull request by frezbo**
Enhances #299 

Signed-off-by: Noel Georgi <git@frezbo.com>

### CODE REVIEW

**Comment by frezbo** [comment]
@mtibben @teeberg This should be good to go

**Comment by teeberg** [comment]
yeaaah, really cool :)) thanks for getting on that so quickly!!

**Comment by teeberg** [comment]
Any chance to get this merged?

**Comment by mtibben** [comment]
I was hoping for a review from @lox as he's most familiar with that code

File: cli/exec.go [hunk 1]
```diff
@@ -17,17 +19,26 @@ import (
...
+	RoleDuration     time.Duration
+	MfaToken         string
+	MfaPrompt        prompt.PromptFunc
+	StartServer      bool
+	CredentialHelper bool
+	Signals          chan os.Signal
+	NoSession        bool
+}
+
+type AwsCredentialHelperData struct {
```
Comment (lox) [inline cli/exec.go]: Could we get a code comment explaining what this is for future generations?

File: cli/exec.go [hunk 2]
```diff
@@ -53,6 +64,10 @@ func ConfigureExecCommand(app *kingpin.Application) {
 		Short('m').
 		StringVar(&input.MfaToken)
 
+	cmd.Flag("json", "AWS credential helper").
```
Comment (lox) [inline cli/exec.go]: https://www.google.com/search?hl=en&q=%22aws%20credential%20helper%22 yields nothing of use, might there be a clearer way to describe this?

**Comment by lox** [comment]
This looks like super straight forward code, but would be good to help folks figure out it's there and how it works. Perhaps also a README addition @teeberg?

**Comment by teeberg** [comment]
Is there anything still missing here?

**Comment by lox** [comment]
I see a lot of whitespace changes in here, is that an issue with your `gofmt` @frezbo or is there an issue with the file?

**Comment by lox** [comment]
I'm going to merge this and give it a cleanup in master. Thnaks @frezbo!

**Comment by frezbo** [comment]
@pecigonzalo this:

```
[profile product]
region=us-east-1
credential_process=aws-vault exec -j product
```

**Comment by teeberg** [comment]
and then instead of manually calling `aws-vault`, you only need to make sure that you have `AWS_PROFILE=product` in your environment, e.g. with the above config, call
```
AWS_PROFILE=product aws sts get-caller-identity
```

**Comment by pecigonzalo** [comment]
Yeah, that is what I did, but I get access denied. I'll put my config in a
moment, as I'm using roles and MFA, it might be that it gets in the way,
altho I tried a few combos.

**Comment by lox** [comment]
Agreed some more docs would be really helpful for folks!

**Comment by frezbo** [comment]
@pecigonzalo I wonder if its the periods `.` in your profile names, can you try using `-`. BTW what does `aws-vault exec -j main.root` return, does that ask for any prompts?

**Comment by StevenACoffman** [comment]
I can successfully use the credential_process for a credentialed source role, but not to use that source role to assume another role. I created a shim bash script to supply my MFA to rule that out as an issue.

For instance I get this error:
```
$ AWS_PROFILE=engineer /usr/local/bin/aws sts get-caller-identity

The source_profile "personal" must specify either static credentials or an assume role configuration
```
This is with an `.aws/config` like:
```
[profile personal]
region = us-east-1
output = json
mfa_serial = arn:aws:iam::123456:mfa/scoffman
credential_process = /Users/scoffman/bin/aws-credential-helper.sh

[profile engineer]
role_arn = arn:aws:iam::123456:role/DevelopmentV1
source_profile = personal
mfa_serial = arn:aws:iam::123456:mfa/scoffman
region = us-east-1
output = json
```
The shim bash script is just this:
```
#!/bin/bash
export TOTP="$(2fa ${AWS_MFA_NAME:-personal})"
if [[ -n "${TOTP:-}" ]]
then
  aws-vault exec --mfa-token=${TOTP} -j personal
else
    echo "No MFA TOTP! 2fa did not find a MFA TOTP."
fi
```

Currently, for assuming a role that has a source_profile attribute using CLI, the source_profile has to have a static credential or it is a profile that is configured to use assume-role [1]. Thus, the credential from credential_process cannot be used since they are external process and considered floating variables. 

[1] https://docs.aws.amazon.com/cli/latest/topic/config-vars.html

**Comment by frezbo** [comment]
@StevenACoffman yeh that makes sense, I mostly use `credential_process` with terraform where the same code deploy stuff to multiple accounts based on the profile name. For profiles that need MFA auth I first do `aws-vault exec <profile> -- bash` and type in the MFA, so that the actual tool using the profile does not prompt for MFA.

**Comment by StevenACoffman** [comment]
This appears to work:
$ AWS_PROFILE=helper /usr/local/bin/aws sts get-caller-identity

+ The source_profile `personal` must specify static credentials.
+ The assumed role profile `engineer` must specify a role_arn, and source_profile `engineer`
+ The credential_process profile `helper` must specify aws-vault  profile `engineer`
```
This is with an `.aws/config` like:
```
[profile personal]
region = us-east-1
output = json
mfa_serial = arn:aws:iam::123456:mfa/scoffman

[profile engineer]
role_arn = arn:aws:iam::123456:role/DevelopmentV1
source_profile = personal
mfa_serial = arn:aws:iam::123456:mfa/scoffman
region = us-east-1
output = json

[profile helper]
credential_process = /Users/scoffman/bin/aws-credential-helper.sh
mfa_serial = arn:aws:iam::594813696195:mfa/scoffman
region = us-east-1
output = json
```
The `aws-credential-helper.sh` shim bash script is just this:
```
#!/bin/bash
export TOTP="$(2fa ${AWS_MFA_NAME:-engineer})"
if [[ -n "${TOTP:-}" ]]
then
  aws-vault exec --mfa-token=${TOTP} -j engineer
else
    echo "No MFA TOTP! 2fa did not find a MFA TOTP."
fi
```
I do not think the shim that supplies MFA is necessary, but haven't tested it.

**Comment by pecigonzalo** [comment]
BTW, I tried without the `.` but it did not work. 
It make sense tho, as the `aws-vault` profile is also using config form the `aws/config` file, like which profile to assume etc. I guess we just need to document how/when to use this to avoid confusion.

**Comment by StevenACoffman** [comment]
@pecigonzalo 
Try:
```
[profile main_root]
region=eu-central-1
credential_process=aws-vault exec -j main_root

[profile dev_admin] 
source_profile=main_root
region=eu-central-1
mfa_serial=arn:aws:iam::123123123:mfa/gonzalo.peci
role_arn=arn:aws:iam::123123123:role/main-root-admin 
output=json

[profile dev_admin_helper] 
region=eu-central-1
mfa_serial=arn:aws:iam::123123123:mfa/gonzalo.peci
output=json
credential_process=aws-vault exec -j dev_admin
```

**Comment by eikeon** [comment]
@StevenACoffman That works for me except it's eating the prompt. For example if I hit enter when it hangs (I didn't see the prompt) then I see: "Error when retrieving credentials from custom-process: Enter token for arn:xxxxxxxxx: "

**Comment by eikeon** [comment]
Ah, I see there's a --prompt

`aws-vault exec -j --prompt=osascript my_profile_name` works for me

**Comment by mhumeSF** [comment]
So for this to work we need two profiles for every assume_role profile?
- 1 for `assume_profile`
- 1 for `helper`

```
[profile default]
credential_process=aws-vault exec -j default

[profile dev]
source_profile = default
role_arn = arn:aws:iam::222222222222:role/dev
mfa_serial = arn:aws:iam::111111111111:mfa/me@example.com

[profile dev_helper]
mfa_serial = arn:aws:iam::111111111111:mfa/me@example.com
credential_process=aws-vault exec -j --prompt=osascript dev
```

**Comment by StevenACoffman** [comment]
That's not how I have it working. 
I thought aws-vault needs to have the credentials stored in a non-default profile.
e.g. `aws-vault add source --keychain login` and then have the config more like this:
```
[default]
credential_process = aws-vault exec -j dev
region = us-east-1
output = json
 
[profile source]
region = us-east-1
mfa_serial = arn:aws:iam::111111111111:mfa/me
 
[profile dev]
mfa_serial = arn:aws:iam::111111111111:mfa/me
region = us-east-1
role_arn = arn:aws:iam::111111111111:role/dev
source_profile = source
```
`default` depends on `dev` which depends on `source`.

**Comment by lox** [comment]
> I thought aws-vault needs to have the credentials stored in a non-default profile.

Hmmm, interesting, I can't think of a reason why creds couldn't be stored in the default, that said, I've never done that. I have `source` profiles too for credentials.

**Comment by glenjamin** [comment]
Could this be added to the main documentation somewhere? This looks like just what I was looking for but I only found it via a colleague

**Comment by mhumeSF** [comment]
I was hung up on this so adding my notes here for anyone else. 
- Set `AWS_SDK_LOAD_CONFIG=true` in environment variables.
- Static credentials are added via `aws-vault add my-aws-account`

```
# ~/.aws/config
[profile my-aws-account]

[profile _source_blue]
source_profile=my-aws-account
role_arn=arn:aws:iam::111111111111:role/blue
mfa_serial=arn:aws:iam::000000000000:mfa/me@example.com

[profile blue]
credential_process=aws-vault exec _source_blue --json
```

This allows one to switch profiles by just setting the `AWS_PROFILE` variable. And now I can add `AWS_PROFILE=blue` in a `.envrc` and when I cd into that directory I am prompted for keychain and mfa; but only if no session is available or expired.

**Comment by mtibben** [comment]
Yes absolutely that would be great @glenjamin, please feel free to submit a PR

**Pull request merged**